core/src/Kokkos_Serial.hpp

```diff
@@ -155,6 +186,13 @@ class Serial {
   uint32_t impl_instance_id() const noexcept { return 0; }
 
   static const char* name();
+
+  inline Impl::SerialInternal* impl_internal_space_instance() const {
+    return m_space_instance.get();
+  }
+
+ private:
+  std::shared_ptr<Impl::SerialInternal> m_space_instance;
```

Comment:
Using a shared pointer makes little sense if we don't allow the creation of non default execution instances.

Reply:
I need a copy-constructor and Kokkos assumes that the copy will have the same `m_space_instance`.

Reply:
I am not following.  Why can't you just return directly a reference to the singleton in `impl_internal_space_instance`?

Reply:
The object likely needs to be copyable to the device. So using the Kokkos HostSharedPtr is what we need here.
Though I agree this needs to be a tracked ptr since the guy owns stuff like scratch memory, and if you try to use these guys in multiple external threads it needs to be tracked.